File: src/main/java/at/yawk/fiction/impl/fanfiction/SearchPageParser.java
package at.yawk.fiction.impl.fanfiction;

import at.yawk.fiction.Image;
import at.yawk.fiction.Pageable;
import at.yawk.fiction.RawText;
import at.yawk.fiction.impl.PageParser;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * @author yawkat
 */
class SearchPageParser extends PageParser<Pageable.Page<FfnStory>> {
    @Override
    protected void parse(Element root, Pageable.Page<FfnStory> page) throws Exception {
        List<FfnStory> stories = new ArrayList<>();

        Elements storyContainers = root.select(".z-list");
        for (Element storyContainer : storyContainers) {
            Element stitle = storyContainer.select("a.stitle").first();

            FfnStory story = new FfnStory();
            story.setId(Integer.parseInt(extractGroup(stitle.attr("href"), "/s/(\\d+)/.*")));
            story.setTitle(stitle.text());
            Element image = stitle.select("img").first();
            if (image.hasAttr("data-original")) {
                story.setImage(new Image(new URI(image.absUrl("data-original"))));
            }
            story.setUri(new URI("https://fanfiction.net/s/" + story.getId() + "/1"));

            for (Element link : storyContainer.select("> a")) {
                String authorIdString = extractGroup(link.attr("href"), "/u/(\\d+)/.*");
                if (authorIdString != null) {
                    FfnAuthor author = new FfnAuthor();
                    author.setId(Integer.parseInt(authorIdString));
                    author.setName(link.text());
                    story.setAuthor(author);
                    break;
                }
            }

            Element descriptionTag = storyContainer.select(".z-indent").first();

            RawText description = new RawText();
            description.setText(descriptionTag.ownText());
            story.setDescription(description);

            getParser(TagParser.class).parse(descriptionTag.select(".xgray").first(), story);

            stories.add(story);
        }

        page.setEntries(stories);

        int maxPage = 0;
        for (Element pager : root.select("#content_wrapper_inner center a")) {
            maxPage = Math.max(maxPage, Integer.parseInt(extractGroup(pager.attr("href"), ".*p=(\\d+)")));
        }
        page.setPageCount(maxPage);
    }

    @Override
    protected Pageable.Page<FfnStory> create() {
        return new Pageable.Page<>();
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fanfiction/SearchPageParser.kt
package at.yawk.fiction.impl.fanfiction

import at.yawk.fiction.*
import at.yawk.fiction.impl.PageParser
import org.jsoup.nodes.Element
import java.net.URI
import java.util.*

/**
 * @author yawkat
 */
internal object SearchPageParser : PageParser<Pageable.Page<Story>>() {
    // yes, Referrer is misspelled in HTTP.
    private val FANFICTION_REFERER = URI.create("https://www.fanfiction.net")

    override fun parse(root: Element): Pageable.Page<Story> {
        val stories = ArrayList<Story>()

        val storyContainers = root.select(".z-list")
        for (storyContainer in storyContainers) {
            val stitle = storyContainer.select("a.stitle").first()

            val id = Integer.parseInt(extractGroup(stitle.attr("href"), "/s/(\\d+)/.*"))
            val title = stitle.text()
            val imageTag = stitle.select("img").first()

            var image: Image? = null
            if (imageTag.hasAttr("data-original")) {
                val imageUrl = imageTag.absUrl("data-original")
                if (imageUrl.endsWith("/75/")) {
                    image = Image(
                            // full-size image
                            imageUrl = URI(imageUrl.substring(0, imageUrl.length - 4) + "/180/"),
                            // thumbnail
                            thumbnailUrl = URI(imageUrl),
                            referer = FANFICTION_REFERER)
                }
            }

            var author: Author? = null
            for (link in storyContainer.select("> a")) {
                val authorIdString = extractGroup(link.attr("href"), "/u/(\\d+)/.*")
                if (authorIdString != null) {
                    author = Author(
                            name = link.text(),
                            providerData = FfnAuthorProviderData(
                                    id = authorIdString.toInt()
                            )
                    )
                    break
                }
            }

            val descriptionTag = storyContainer.select(".z-indent").first()

            val description = RawText(descriptionTag.ownText())

            var story = Story(
                    title = title,
                    uri = URI("https://fanfiction.net/s/$id/1"),
                    image = image,
                    description = description,
                    author = author,
                    providerData = FfnStoryProviderData(
                            id = id
                    )
            ).merge(TagParser.parse(descriptionTag.select(".xgray").first()))
            stories.add(story)
        }


        var maxPage = 0
        for (pager in root.select("#content_wrapper_inner center a")) {
            maxPage = Math.max(maxPage, Integer.parseInt(extractGroup(pager.attr("href"), ".*p=(\\d+)")))
        }

        return Pageable.Page(stories, maxPage, false)
    }
}
